base/src/main/java/io/spine/code/proto/FieldDeclaration.java

```diff
@@ -20,39 +20,117 @@
+     *     <li>Named {@code id} or the name ends with {@code _id}.</li>
+     *     <li>Declared inside an {@linkplain EntityOption#getKind() entity state message}.
+     * </ul>
+     *
+     * @return {@code true} if the field is an entity ID, {@code false} otherwise
+     */
+    public boolean isEntityId() {
+        return isFirstField() && matchesIdName() && isEntityField();
     }
 
-    public FieldName name() {
-        return FieldName.of(descriptor);
+    /**
+     * Determines whether the field is a command ID.
+     *
+     * <p>A command ID is the first field of a message declared in a
+     * {@link io.spine.base.CommandMessage.File command file}.
+     *
+     * @return {@code true} if the field is a command ID, {@code false} otherwise
+     */
+    public boolean isCommandId() {
+        return isFirstField() && isCommandsFile();
     }
 
-    public FieldDescriptorProto descriptor() {
-        return descriptor;
+    /**
+     * Determines whether the declaration is a scalar type.
+     *
+     * @return {@code true} if the declaration neither map nor repeated, {@code false} otherwise
+     */
+    public boolean isScalar() {
+        return !isMap() && !isRepeated();
     }
 
     /**
-     * Obtains comments going before the field.
+     * Determines whether the declaration is not a scalar type.
      *
-     * @return the leading field comments or {@code Optional.empty()} if there are no comments
+     * @return {@code true} if the declaration either map or repeated, {@code false} otherwise
      */
-    public Optional<String> leadingComments() {
-        return originMessage.documentation()
-                            .getFieldLeadingComments(descriptor);
+    public boolean isNotScalar() {
+        return isMap() || isRepeated();
+    }
+
+    public boolean isRepeated() {
+        return FieldTypes.isRepeated(field);
+    }
+
+    public boolean isMap() {
+        return FieldTypes.isMap(field);
+    }
+
+    /** The {@link FieldDescriptor.JavaType JavaType} of the declaration. */
+    public FieldDescriptor.JavaType javaType() {
+        return field.getJavaType();
+    }
+
+    /** Obtains the descriptor of the value of a map. */
+    public FieldDeclaration valueDeclaration() {
+        FieldDescriptor valueDescriptor = FieldTypes.valueDescriptor(field);
+        return new FieldDeclaration(valueDescriptor);
+    }
+
+    private boolean isEntityField() {
+        EntityOption entityOption = field.getContainingType()
+                                         .getOptions()
+                                         .getExtension(OptionsProto.entity);
+        EntityOption.Kind entityKind = entityOption.getKind();
+        return entityKind.getNumber() > 0;
+    }
+
+    private boolean matchesIdName() {
+        String name = field.getName();
+        return "id".equals(name) || name.endsWith("_id");
+    }
+
+    private boolean isFirstField() {
+        return field.getIndex() == 0;
```

Comment:
Shouldn't the field index be `1`?

Reply:
The index is zero-based. It is **not** about the number in the Proto declaration (`string field = 1`).